test/infra/cchost.py

```diff
@@ -0,0 +1,414 @@
+        self.workspace.joinpath("member0_enc_privk.pem").write_text(
+            self.encryption_private_key
+        )
+        self.workspace.joinpath("member0_enc_pubk.pem").write_text(
+            encryption_public_key
+        )
+
+        self.thread = threading.Thread(target=self._execute)
+
+        self.cond = threading.Condition()
+        self.ready = False
+        self.thread_exception = None
+
+        self.loop = asyncio.new_event_loop()
+
+        # Event objects must be created on the loop they will be accessed from.
+        async def create_event():
+            return asyncio.Event()
+
+        self.shutdown_request = self.loop.run_until_complete(create_event())
+
+    def __enter__(self):
+        self.thread.start()
+
+        try:
+            with self.cond:
+                while not self.ready:
+                    self.cond.wait()
+                    if self.thread_exception is not None:
+                        raise self.thread_exception
+
+        except:
+            # This is to handle the case where the wait is interrupted,
+            # by a KeyboardInterrupt for example.
+            self.loop.call_soon_threadsafe(self.shutdown_request.set)
+            self.thread.join()
+            raise
+
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        try:
+            self.loop.call_soon_threadsafe(self.shutdown_request.set)
+        finally:
+            self.thread.join()
+
+        if self.thread_exception is not None:
+            raise self.thread_exception
+
+    def _execute(self):
+        """
+        Entrypoint for the background thread.
+        """
+        try:
+            self.loop.run_until_complete(self._start_service())
+        except BaseException as e:
+            with self.cond:
+                self.thread_exception = e
+                self.cond.notify_all()
+
+    async def _start_service(self):
+        """
+        Run cchost in a loop, handling shutdown requests.
+        """
+        self._populate_workspace()
+
+        try:
+            await self._start_process()
+        except aiotools.TaskGroupError as e:
+            if match_taskgroup_error(e, ShutdownRequestException):
+                return
+            else:
+                raise
+
+    async def _start_process(self):
+        process = await asyncio.create_subprocess_exec(
+            self.binary,
+            "--config",
+            self.workspace / "config.json",
+            cwd=self.workspace,
+            start_new_session=True,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env={
+                **os.environ,
+                # This is needed for the attested fetch script to know where
+                # to contact the CCF service. The script is started by cchost,
+                # so it inherits any of cchost's environment variables.
+                "CCF_SERVICE_PORT": str(self.rpc_port),
```

Comment:
Even the port number isn't guaranteed to be the same. It's quite possible that cchost inside a docker image is listening on port 8000 but that is exposed as port 443 to the outside world.

Reply:
Ah you're right. Hm, ok, let's leave it for now until we find a better solution.

Reply:
Heh, so this is now broken by the random port assignment, since we don't have a way of setting an environment variable with the value ahead of starting the process. I need to figure out if there's a way to get this port number from within the application to pass it as a callback URL.